<a href="https://colab.research.google.com/github/douglas-t/acs-community-deployment/blob/master/TP1_MLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# 1 Charger les bibliothèques essentielles
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:
from google.colab import drive
import os
drive.mount('/content/drive')
os.chdir('/content/drive/My Drive/datasets')

Mounted at /content/drive


In [5]:
#2. Charger votre dataset diabets.csv dans la dataframe datadiabetes :
data_diabetes = pd.read_csv('2021_diabetes.csv',sep=',',index_col=None,header=0,on_bad_lines='skip')

In [6]:
#test de chargement de données
data_diabetes.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               2000 non-null   int64  
 1   Glucose                   2000 non-null   int64  
 2   BloodPressure             2000 non-null   int64  
 3   SkinThickness             2000 non-null   int64  
 4   Insulin                   2000 non-null   int64  
 5   BMI                       2000 non-null   float64
 6   DiabetesPedigreeFunction  2000 non-null   float64
 7   Age                       2000 non-null   int64  
 8   Outcome                   2000 non-null   int64  
dtypes: float64(2), int64(7)
memory usage: 140.8 KB


In [8]:
data_diabetes.describe()

Pregnancies      Glucose  BloodPressure  SkinThickness      Insulin  \
count  2000.000000  2000.000000    2000.000000    2000.000000  2000.000000   
mean      3.703500   121.182500      69.145500      20.935000    80.254000   
std       3.306063    32.068636      19.188315      16.103243   111.180534   
min       0.000000     0.000000       0.000000       0.000000     0.000000   
25%       1.000000    99.000000      63.500000       0.000000     0.000000   
50%       3.000000   117.000000      72.000000      23.000000    40.000000   
75%       6.000000   141.000000      80.000000      32.000000   130.000000   
max      17.000000   199.000000     122.000000     110.000000   744.000000   

               BMI  DiabetesPedigreeFunction          Age      Outcome  
count  2000.000000               2000.000000  2000.000000  2000.000000  
mean     32.193000                  0.470930    33.090500     0.342000  
std       8.149901                  0.323553    11.786423     0.474498  
min       0.000000                  0.078000    21.000000     0.000000  
25%      27.375000                  0.244000    24.000000     0.000000  
50%      32.300000                  0.376000    29.000000     0.000000  
75%      36.800000                  0.624000    40.000000     1.000000  
max      80.600000                  2.420000    81.000000     1.000000

**1. Préparation du dataset**

In [10]:
#Séparer les variables X et cible X
X=data_diabetes.drop('Outcome',axis=1)
Y=data_diabetes['Outcome']

In [11]:
#Séparer en train/test
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,random_state=42)

In [12]:
#Normaliser les données
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)

**2. MLP avec Scikit-Learn **

In [13]:
#Créer un perceptron multicouche avec une ou deux couches cachées
from sklearn.neural_network import MLPClassifier
## Définir le MLP
mlp=MLPClassifier(hidden_layer_sizes=(16,8),activation="relu", solver="adam",max_iter=2000,random_state=42)
mlp.fit(X_train,Y_train)

MLPClassifier(hidden_layer_sizes=(16, 8), max_iter=2000, random_state=42)

In [15]:
#Entraîner le modèle sur les données normalisées
mlp.fit(X_train,Y_train)
#prediction
Y_pred=mlp.predict(X_test)

In [16]:
#Afficher la précision sur l’ensemble de test
from sklearn.metrics import accuracy_score,classification_report
print("Précision sur l'ensemble de test : ",accuracy_score(Y_pred,Y_test))
print(classification_report(Y_test,Y_pred))

Précision sur l'ensemble de test :  0.74
              precision    recall  f1-score   support

           0       0.74      0.90      0.81       253
           1       0.73      0.47      0.57       147

    accuracy                           0.74       400
   macro avg       0.74      0.68      0.69       400
weighted avg       0.74      0.74      0.72       400



In [17]:
#MLP avec TensorFlow
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense,Dropout

In [18]:

print(X_train_scaled.shape[1])

8


In [22]:
#construire le model
model=Sequential()
model.add(Dense(16,activation='relu',input_shape=(X_train_scaled.shape[1],)))
model.add(Dense(8,activation='relu'))
model.add(Dense(1,activation='sigmoid'))


In [23]:
#Compiler le modèle
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [26]:
# Entraîner le modèle et visualiser l’évolution des métriques
history=model.fit(X_train_scaled,Y_train,epochs=100,batch_size=32,validation_split=0.1)

Epoch 1/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8958 - loss: 0.2583 - val_accuracy: 0.7812 - val_loss: 0.4054
Epoch 2/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8937 - loss: 0.2632 - val_accuracy: 0.7688 - val_loss: 0.4036
Epoch 3/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8870 - loss: 0.2653 - val_accuracy: 0.7625 - val_loss: 0.4030
Epoch 4/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8901 - loss: 0.2792 - val_accuracy: 0.7688 - val_loss: 0.4027
Epoch 5/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8927 - loss: 0.2598 - val_accuracy: 0.7875 - val_loss: 0.4031
Epoch 6/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8943 - loss: 0.2464 - val_accuracy: 0.7688 - val_loss: 0.3995
Epoch 7/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8915 - loss: 0.2608 - val_accuracy: 0.7875 - val_loss: 0.4004
Epoch 8/50
45/45 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8905 - loss: 0.2646 - val_accuracy: 0.7875 - val_loss:

In [27]:
#Evaluer le modèle
loss,accuracy=model.evaluate(X_test_scaled,Y_test)
print("Précision sur l'ensemble de test : ",accuracy)

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8239 - loss: 0.3955 
Précision sur l'ensemble de test :  0.8475000262260437


In [ ]:
#Comparer les deux modèles


**Précision globale (Accuracy)**

*   Modèle Scikit-learn : Précision sur l'ensemble de test = 0.74 (74 %).
*   Modèle TensorFlow : Précision sur l'ensemble de test = 0.8475 (84.75 %).

**Observation :** Le modèle TensorFlow est nettement plus performant en termes de précision globale, avec une amélioration d’environ 10.75 points de pourcentage (0.8475 - 0.74). Cela suggère que le modèle TensorFlow est mieux capable de classer correctement les échantillons de l'ensemble de test.